### ISSUE TITLE
invalid-assignment false positive: dict literal not narrowed to Mapping[Literal[...], ...] inside union

### ISSUE THREAD

**Issue by eitan-protego** [astral-sh/ty#3487]
### Summary

A dict literal like `{"$in": ["a", "b"]}` is correctly assignable to `Mapping[Literal["$in", "$nin"], Sequence[str]]` when that type is the direct target. However, when the target type is a **union** that includes that `Mapping` type, ty fails to narrow the dict literal's key type from `str` to `Literal["$in"]` and rejects the assignment.

This blocks a common pattern for typed MongoDB query operators where an `Op[T]` type is defined as a union of the raw value type and several operator Mapping types.

### Repro

```python
from typing import Literal
from collections.abc import Mapping, Sequence

type SelfOp[T] = Mapping[Literal["$eq", "$ne"], T]
type ListOp[T] = Mapping[Literal["$in", "$nin"], Sequence[T]]
type Op[T] = T | SelfOp[T] | ListOp[T]

# OK — direct assignment to the specific Mapping type:
direct: ListOp[str] = {"$in": ["a", "b"]}

# Error — assignment to the union that includes that Mapping:
union: Op[str] = {"$in": ["a", "b"]}
#                ^^^^^^^^^^^^^^^^^^^
# error[invalid-assignment]: Object of type `dict[str, list[str]]`
#                            is not assignable to `Op[str]`
```

[ty playground](https://play.ty.dev/59909aa8-c00a-42cd-8bc7-3909e62f6f0a)

### Expected behavior

Since `{"$in": ["a", "b"]}` matches `ListOp[str]`, and `ListOp[str]` is a member of `Op[str]`, the assignment should be accepted. The dict literal's key `"$in"` should be narrowed to `Literal["$in"]` as pyright and basedpyright do.

### Other type checkers

| Checker | Result |
|---|---|
| pyright 1.1.409 | pass |
| basedpyright 1.39.3 | pass |
| **mypy 2.0.0** | **error** (same issue) |
| pyrefly 0.64.1 | pass |
| **ty 0.0.38** | **error** |

Note: mypy also rejects this, so this is not universally agreed upon.

### Related

- #2103 — dict with known literal keys assignable to `Mapping[Literal, ...]` (fixed, but only for direct assignment — not through a union)

### Version

ty 0.0.34, confirmed still present in 0.0.35, 0.0.36, 0.0.37, and 0.0.38

### Version

ty 0.0.29 (438a78d68 2026-04-05)
